File: typescript/src/algorithms/binary-search.ts
/**
 * Binary search implementation.
 * Time: O(log n)
 * Space: O(1)
 */

type Comparator<T> = (a: T, b: T) => number;

export function binarySearch<T>(arr: T[], target: T, comparator?: Comparator<T>): number | undefined {
  const compare = comparator || ((a: T, b: T) => {
    if (a < b) return -1;
    if (a > b) return 1;
    return 0;
  });

  let left = 0;
  let right = arr.length - 1;

  while (left <= right) {
    const mid = left + Math.floor((right - left) / 2);
    const cmp = compare(arr[mid], target);

    if (cmp === 0) {
      return mid;
    } else if (cmp < 0) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return undefined;
}

export function lowerBound<T>(arr: T[], target: T, comparator?: Comparator<T>): number {
  const compare = comparator || ((a: T, b: T) => {
    if (a < b) return -1;
    if (a > b) return 1;
    return 0;
  });

  let left = 0;
  let right = arr.length;

  while (left < right) {
    const mid = left + Math.floor((right - left) / 2);
    if (compare(arr[mid], target) < 0) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }

  return left;
}

export function upperBound<T>(arr: T[], target: T, comparator?: Comparator<T>): number {
  const compare = comparator || ((a: T, b: T) => {
    if (a < b) return -1;
    if (a > b) return 1;
    return 0;
  });

  let left = 0;
  let right = arr.length;

  while (left < right) {
    const mid = left + Math.floor((right - left) / 2);
    if (compare(arr[mid], target) <= 0) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }

  return left;
}

File: csharp/src/Algorithms/BinarySearch.cs
namespace DSA.Algorithms;

/// <summary>
/// Binary search implementation.
/// Time: O(log n)
/// Space: O(1)
/// </summary>
public static class BinarySearch
{
    public static int? Search<T>(IList<T> arr, T target, IComparer<T>? comparer = null)
    {
        comparer ??= Comparer<T>.Default;

        int left = 0;
        int right = arr.Count - 1;

        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            int cmp = comparer.Compare(arr[mid], target);

            if (cmp == 0)
                return mid;
            if (cmp < 0)
                left = mid + 1;
            else
                right = mid - 1;
        }

        return null;
    }

    public static int LowerBound<T>(IList<T> arr, T target, IComparer<T>? comparer = null)
    {
        comparer ??= Comparer<T>.Default;

        int left = 0;
        int right = arr.Count;

        while (left < right)
        {
            int mid = left + (right - left) / 2;
            if (comparer.Compare(arr[mid], target) < 0)
                left = mid + 1;
            else
                right = mid;
        }

        return left;
    }

    public static int UpperBound<T>(IList<T> arr, T target, IComparer<T>? comparer = null)
    {
        comparer ??= Comparer<T>.Default;

        int left = 0;
        int right = arr.Count;

        while (left < right)
        {
            int mid = left + (right - left) / 2;
            if (comparer.Compare(arr[mid], target) <= 0)
                left = mid + 1;
            else
                right = mid;
        }

        return left;
    }
}
